File: runtime/java/units/wasi/fd_filestat_set_size.java
// Truncate or extend a file to `size`.
// FileChannel.truncate only shrinks, so growing is done by writing one zero byte at size-1 (the gap is a sparse hole), matching ftruncate's grow semantics.
int wasi_fd_filestat_set_size(int fd, long size) {
    Object e = fds.get(fd);
    if (!(e instanceof Handle)) {
        return WASI_BADF;
    }
    if (lacksRight(fd, R_FD_FILESTAT_SET_SIZE)) {
        return WASI_NOTCAPABLE;
    }
    java.nio.channels.FileChannel ch = ((Handle) e).ch;
    try {
        long cur = ch.size();
        if (size < cur) {
            ch.truncate(size);
        } else if (size > cur) {
            ch.write(java.nio.ByteBuffer.wrap(new byte[] {0}), size - 1);
        }
    } catch (java.io.IOException ex) {
        return WASI_IO;
    }
    return WASI_OK;
}

File: runtime/go/units/wasi/fd_filestat_set_size.go
func (w *WASI) wasi_fd_filestat_set_size(fd uint32, size uint64) uint32 {
    f, ok := w.fds[fd].(*os.File)
    if !ok {
        return wasiBadf
    }
    if e := w.checkRight(fd, rightFdFilestatSetSize); e != wasiOk {
        return e
    }
    if err := f.Truncate(int64(size)); err != nil {
        return wasiIo
    }
    return wasiOk
}
